metrics/metricshttpx/middleware.go

```diff
@@ -0,0 +1,68 @@
+package metricshttpx
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"golang.org/x/net/context"
+
+	"github.com/remind101/pkg/httpx"
+	"github.com/remind101/pkg/httpx/middleware"
+	"github.com/remind101/pkg/metrics"
+)
+
+// ResponseTimeReporter reports timing metrics using metrics package
+//
+// Usage:
+//   r := httpx.NewRouter()
+//   ...
+//   r.HandleFunc("/", func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+//   	w.WriteHeader(http.StatusCreated)
+//   	return nil
+//   }).Methods("GET")
+//   s := NewResponseTimeReporter(r, r)
+//
+func NewResponseTimeReporter(handler httpx.Handler, router *httpx.Router) *responseTimeReporter {
+	if router == nil {
+		panic("NewResponseTimeReporter: router is requred")
+	}
+	return &responseTimeReporter{handler, router}
+}
+
+type responseTimeReporter struct {
+	handler httpx.Handler
+	router  *httpx.Router
+}
+
+func (h *responseTimeReporter) ServeHTTPContext(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+	t := metrics.ResponseTime()
+	defer t.Done()
+
+	rw := middleware.NewResponseWriter(w) // exposes status code
+	err := h.handler.ServeHTTPContext(ctx, rw, r)
+
+	route := fmt.Sprintf("%s %s", r.Method, templatePath(h.router, r))
+	status := strconv.Itoa(rw.Status())
+	t.SetTags(map[string]string{
+		"route":  route,
+		"status": status,
+	})
+
+	return err
+}
+
+func templatePath(router *httpx.Router, r *http.Request) string {
```

Comment:
copied this from httpx/middleware/newrelic. I'll extract when I get to middlewares for frameworks other than martini/httpx.